importer/src/queue.ts

```diff
@@ -0,0 +1,158 @@
+ */
+
+
+import {Buffer} from 'node:buffer';
+import {type ParsedEntity} from './parser.ts';
+import amqp from 'amqplib';
+import log from './helpers/logger.ts';
+
+
+/** Queue which stores parsed entities that have to be imported into the BookBrainz database. */
+export class ImportQueue {
+	constructor({
+		connectionUrl = 'amqp://localhost',
+		isPersistent = true,
+		prefetchLimit = 5,
+		queueName = 'bookbrainz-import',
+		failureQueue = false
+	}: Partial<ImportQueueOptions> = {}) {
+		this.connectionUrl = connectionUrl;
+		this.isPersistent = isPersistent;
+		this.prefetchLimit = prefetchLimit;
+		this.queueName = queueName;
+		if (failureQueue) {
+			this.failureQueueName = failureQueue;
+		}
+	}
+
+	/**
+	 * Opens the connection to the AMQP server and creates a message channel.
+	 * Ensures that our persistent import queue exists and creates it if necessary.
+	 */
+	async open() {
+		this.connection = await amqp.connect(this.connectionUrl);
+		this.channel = await this.connection.createChannel();
+		await this.channel.prefetch(this.prefetchLimit);
+
+		if (this.failureQueueName) {
+			this.channel.assertQueue(this.failureQueueName, {durable: this.isPersistent});
+		}
+
+		return this.channel.assertQueue(this.queueName, {durable: this.isPersistent});
+	}
+
+	/** Closes the connection to the AMQP server. */
+	async close(): Promise<boolean> {
+		// wait until the channel closes to guarantee that all sent messages went through
+		await this.channel?.close();
+		await this.connection?.close();
+
+		return Boolean(this.connection);
+	}
+
+	/** Appends the given entity to the import queue. */
+	push(entity: QueuedEntity): boolean {
+		let message: string;
+		try {
+			message = JSON.stringify(entity);
+		}
+		catch (error) {
+			log.error(`Failed to serialize entity ${entity.originId}: ${error}`);
+			return false;
+		}
+
+		return this.channel.sendToQueue(this.queueName, Buffer.from(message), {
+			persistent: this.isPersistent
+		});
+	}
+
+	/**
+	 * Registers a consumer function for queued entities which is called with already parsed messages.
+	 * Acknowledgement of messages happens automatically according to the returned success value of the consumer.
+	 */
+	onData(consumer: (entity: QueuedEntity) => Promise<boolean>) {
+		return this.channel.consume(this.queueName, async (message) => {
+			let entity: QueuedEntity;
+			try {
+				entity = JSON.parse(message.content.toString());
+			}
+			catch (error) {
+				log.warn('Skipping invalid message:', error.message);
+				// acknowledge invalid message, otherwise it will be requeued
+				this.channel.ack(message);
+				return;
+			}
+
+			const success = await consumer(entity);
+
+			// Acknowledge all consumed messages and put problematic messages into a separate queue (if configured).
+			// We do this to avoid a requeue/redelivery loop which occurs if the consumer never accepts the entity.
+			this.channel.ack(message);
+			if (!success && this.failureQueueName) {
```

Comment:
Would it make sense to log errors here if `!success`, regardless of the presence of a failure queue?

Reply:
Probably not, at this point we only know that the consumer failed but not how. Usually the reason has already been logged by the consumer callback at this point.
We could additionally log them here to be sure we won't miss any discarded message, but honestly the code is already cluttered with too many logger statements for my taste. Most errors are currently logged twice at least, although I had already removed code which first logged an error and then threw it..